content/courses/advanced-ada/chapters/types.rst

```diff
@@ -6,14 +6,432 @@ Types
++                       +-------------------------+---------------------------+
+|                       | :ada:`Wide_Wide_Value`  | :ada:`Wide_Wide_String`   |
++-----------------------+-------------------------+---------------------------+
+
+We discuss more about :ada:`Wide_String` and :ada:`Wide_Wide_String` in
+:ref:`another section <Wide_Wide_Strings>`.
+
+Width attribute
+~~~~~~~~~~~~~~~
+
+When converting a value to a string by using the :ada:`Image` attribute, we get
+a string with variable width. We can assess the maximum width of that string
+for a specific subtype by using the :ada:`Width` attribute. For example,
+:ada:`Integer'Width` gives us the maximum width returned by the :ada:`Image`
+attribute when converting a value of :ada:`Integer` type to a string of
+:ada:`String` type.
+
+This attribute is useful when we're using bounded strings in our code to store
+the string returned by the :ada:`Image` attribute. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Types.Width_Attr
+
+    with Ada.Text_IO;         use Ada.Text_IO;
+    with Ada.Strings;         use Ada.Strings;
+    with Ada.Strings.Bounded;
+
+    procedure Show_Width_Attr is
+       package B_Str is new
+         Ada.Strings.Bounded.Generic_Bounded_Length (Max => Integer'Width);
+       use B_Str;
+
+       Str_I : Bounded_String;
+
+       I : constant Integer := 42;
+       J : constant Integer := 103;
+    begin
+       Str_I := To_Bounded_String (I'Image);
+       Put_Line ("Value:         " & To_String (Str_I));
+       Put_Line ("String Length: " & Length (Str_I)'Image);
+       Put_Line ("----");
+
+       Str_I := To_Bounded_String (J'Image);
+       Put_Line ("Value:         " & To_String (Str_I));
+       Put_Line ("String Length: " & Length (Str_I)'Image);
+    end Show_Width_Attr;
+
+In this example, we're storing the string returned by :ada:`Image` in the
+:ada:`Str_I` variable of :ada:`Bounded_String` type.
+
+Similar to the :ada:`Image` and :ada:`Value` attributes, the :ada:`Width`
+attribute is also available for string types other than :ada:`String`. In fact,
+we can use:
+
+- the :ada:`Wide_Width` attribute for strings returned by :ada:`Wide_Image`;
+  and
+
+- the :ada:`Wide_Wide_Width` attribute for strings returned by
+  :ada:`Wide_Wide_Image`.
+
+Base
+~~~~
+
+The :ada:`Base` attribute gives us the unconstrained version of a subtype. As
+an example, let's say we declared a subtype of the :ada:`Integer` type named
+:ada:`One_To_Ten`:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Types.Base_Attr
+
+    package My_Integers is
+
+       subtype One_To_Ten is Integer range 1 .. 10;
+
+    end My_Integers;
+
+If we then use the :ada:`Base` attribute |mdash| by writing
+:ada:`One_To_Ten'Base` |mdash|, we're actually referring to the
+:ada:`Integer` type (because the :ada:`Integer` type was used as the base
+subtype of the :ada:`One_To_Ten` subtype).
```

Comment:
Subtypes just introduce names for constraints on existing types or subtypes. So whenever we see "One_To_Ten" we are really working with type Integer. The only difference is the values returned for 'First and 'Last. You could say "One_To_Ten'Succ (One_To_Ten'Last)" and assign that value of 11 that to an Integer variable, for example. So, One_To_Ten'Base refers to the same base type as Integer'Base, which is the underlying hardware type representing type Integer but is not Integer itself.

Reply:
Thanks! I reworked this paragraph and tried to integrate your comments.